calc/vxlan_resolver.go

```diff
@@ -0,0 +1,476 @@
+	pendingRouteUpdates.Iter(func(item interface{}) error {
+		r := item.(vxlanRoute)
+		logCxt := logrus.WithField("route", r)
+		if !c.routeReady(r) {
+			return nil
+		}
+		logCxt.Debug("Sending VXLAN route update")
+		c.callbacks.OnRouteUpdate(&proto.RouteUpdate{
+			Type: proto.RouteType_VXLAN,
+			Node: r.node,
+			Dst:  r.dst.String(),
+			Gw:   c.determineGatewayForRoute(r),
+		})
+		return nil
+	})
+}
+
+// withdrawRoute will send a *proto.RouteRemove for the given route.
+func (c *VXLANResolver) withdrawRoute(r vxlanRoute) {
+	logrus.WithField("route", r).Debug("Sending VXLAN route remove")
+	c.callbacks.OnRouteRemove(r.dst.String())
+}
+
+func (c *VXLANResolver) sendVTEPUpdate(node string) bool {
+	logCxt := logrus.WithField("node", node)
+	tunlAddr, ok := c.nodeNameToVXLANTunnelAddr[node]
+	if !ok {
+		logCxt.Debug("Missing vxlan tunnel address for node, cannot send VTEP")
+		return false
+	}
+	parentDeviceIP, ok := c.nodeNameToIPAddr[node]
+	if !ok {
+		logCxt.Debug("Missing IP for node, cannot send VTEP")
+		return false
+	}
+
+	logCxt.Debug("Sending VTEP to dataplane")
+	vtep := &proto.VXLANTunnelEndpointUpdate{
+		Node:           node,
+		ParentDeviceIp: parentDeviceIP,
+		Mac:            c.vtepMACForHost(node),
+		Ipv4Addr:       tunlAddr,
+	}
+	c.callbacks.OnVTEPUpdate(vtep)
+	return true
+}
+
+func (c *VXLANResolver) sendVTEPRemove(node string) {
+	logrus.WithField("node", node).Debug("Withdrawing VTEP from dataplane")
+	c.callbacks.OnVTEPRemove(node)
+}
+
+// routeWithinVXLANPool checks if the provided route is within a configured IP pool with
+// VXLAN enabled.
+func (c *VXLANResolver) routeWithinVXLANPool(r vxlanRoute) bool {
+	for _, pool := range c.vxlanPools {
+		if c.containsRoute(pool, r) {
+			return true
+		}
+	}
+	return false
+}
+
+// routesFromBlock returns a list of routes which should exist based on the provided
+// allocation block. Right now, we only support strict affinity so this will always
+// be the block's CIDR.
+func (c *VXLANResolver) routesFromBlock(blockKey string, b *model.AllocationBlock) map[string]vxlanRoute {
+	if b.Host() == c.hostname {
+		logrus.Debug("Skipping VXLAN routes for local node")
+		return nil
+	}
+
+	r := vxlanRoute{
+		dst:  ip.CIDRFromCalicoNet(b.CIDR),
+		node: b.Host(),
+	}
+	return map[string]vxlanRoute{r.Key(): r}
+}
+
+// determineGatewayForRoute determines which gateway to use for this route. For VXLAN routes, the
+// gateway is the remote node's IPv4VXLANTunnelAddr. If we don't know the remote node's tunnel
+// address, this function will return an empty string.
+func (c *VXLANResolver) determineGatewayForRoute(r vxlanRoute) string {
+	if ip, ok := c.nodeNameToVXLANTunnelAddr[r.node]; ok {
+		return ip
+	}
+	return ""
+}
+
+// vtepMACForHost calculates a deterministic MAC address based on the provided host.
+// The returned address matches the address assigned to the VXLAN device on that node.
+func (c *VXLANResolver) vtepMACForHost(nodename string) string {
```

Comment:
Does this match flannel's algorithm?  It'd probably need to if we want to support migration at some point.

Reply:
Bah, it looks like flannel doesn't use an algorithmic approach for this - it relies on the kernel to generate a MAC for the device when creating it, rather than doing deterministic calculation of it.

https://github.com/coreos/flannel/blob/master/backend/vxlan/vxlan.go#L137

Reply:
@fasaxc wondering if we can merge as-is and then do a follow up to adjust this.

I'm not really sure what we can do about this - we need to teach flannel about our local VXLAN tunnel endpoints, and also learn about Flannel's in order to make this work, which means either augmenting Calico to write to flannel's API or vice-versa, I think.

Reply:
Maybe we just start storing the MAC in the datamodel?  We already store the IP.